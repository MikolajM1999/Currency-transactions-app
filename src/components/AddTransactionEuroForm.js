import React from 'react'
import { useState } from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import { addTransaction } from '../actions/transactions'
import allTransactionsSummary from '../selectors/allTransactionsSummary'

export const AddTransactionEuroForm = (props) => {
   const [error, setError] = useState('')
   const [name, setName] = useState('')
   const [euro, setEuro] = useState('')
   const onInputNameChange = (e) => {
      setName(e.target.value)
   }
   const onInputValueChange = (e) => {
      const value = e.target.value
      if (!value || value.match(/^\d{1,}(\.\d{0,2})?$/)) {
         setEuro(value)
      }
   }
   const onSubmit = (e) => {
      e.preventDefault()
      if (euro === '') {
         setError('Fill in EURO amount at least!')
      } else {
         props.addTransaction(name.trim(), euro)
         setName('')
         setEuro('')
         setError('')
      }
   }
   return (
      <section className="summary-section app__summary-section">
         <div className="transaction-section summary-section__transaction-section">
            <p className="transaction-section__p">Add new transaction:</p>
            <form className="form-add transaction-section__form" onSubmit={onSubmit}>
               <input className="form-add__input" type="text" placeholder="name" value={name} onChange={onInputNameChange} />
               <input className="form-add__input" type="text" placeholder="EURO" value={euro} onChange={onInputValueChange} />
               <button className="form-add__button">Add the transaction</button>
            </form>
            {error && <p className="transaction-section__error">{error}</p>}
            <p className="transaction-section__p">All transactions summary:</p>
            <p className="transaction-section__p">EURO: {Math.round(props.summaryObject.inEURO * 100) / 100}</p>
            <p className="transaction-section__p">PLN: {Math.round(props.summaryObject.inPLN * 100) / 100}</p>
         </div>
      </section>
   )
}

const mapStateToProps = (state) => ({
   summaryObject: allTransactionsSummary(state.transactions, state.currency.euroToPLN)
})

const mapDispatchToProps = (dispatch) => ({
   addTransaction: (name, euro) => dispatch(addTransaction({ name, euro, id: uuid() }))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddTransactionEuroForm)