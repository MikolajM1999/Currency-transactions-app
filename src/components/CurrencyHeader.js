import React from 'react'
import { connect } from 'react-redux'
import { uploadEuroToPLN } from '../actions/currency'

export const CurrencyHeader = (props) => {
   const onInputChange = (e) => {
      const value = e.target.value.trim()
      if (!value || value.match(/^\d{1,}(\.\d{0,9})?$/)) {
         props.uploadEuroToPLN(value)
      }
   }
   return (
      <section className="currency app__currency">
         <div className="currency-div currency__container">
            <h3 className="currency-div__euro-h3">1 EURO = {Math.round(props.euroToPLN * 100) / 100 || 'unknown value of'} PLN</h3>
            <input className="currency-div__input" type="text" placeholder='enter the EURO exchange rate to PLN' onChange={onInputChange} />
            <p className="currency-div__mess">Remember to not type "<span className="span currency-div__span--big-char">,</span>" use "<span className="span currency-div__span--big-char">.</span>" instead!</p>
            <p className="currency-div__mess">Use only numbers!</p>
         </div>
      </section>
   )
}

const mapStateToProps = (state) => ({
   euroToPLN: state.currency.euroToPLN
})

const mapDispatchToProps = (dispatch) => ({
   uploadEuroToPLN: (value) => dispatch(uploadEuroToPLN(value))
})

export default connect(mapStateToProps, mapDispatchToProps)(CurrencyHeader)