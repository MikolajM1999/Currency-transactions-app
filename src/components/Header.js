import React from 'react'

export const Header = () => (
   <header className="header app__header">
      <h1 className="header__title">Currency-Transactions-App</h1>
   </header>
)

export default Header