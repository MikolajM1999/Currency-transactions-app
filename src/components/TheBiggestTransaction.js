import React from 'react'
import { connect } from 'react-redux'
import selectTheBiggestTransaction from '../selectors/selectTheBiggestTransaction'

export const TheBiggestTransaction = ({ theTransaction, euroToPLN }) => (
   <div className="the-biggest-transaction app__most-exensive-transaction-div">
      <p className="the-biggest-transaction__title">The most expensive transaction is:</p>
      {theTransaction && (
         <div className="summary-the-biggest-transaction the-biggest-transaction__div">
            <p className="summary-the-biggest-transaction__p">{theTransaction.name}</p>
            <p className="summary-the-biggest-transaction__p">EURO: {Math.round(theTransaction.euro * 100) / 100}</p>
            <p className="summary-the-biggest-transaction__p">PLN: {Math.round(theTransaction.euro * euroToPLN * 100) / 100}</p>
         </div>
      )}
   </div>
)

const mapStateToProps = (state) => ({
   theTransaction: selectTheBiggestTransaction(state.transactions),
   euroToPLN: state.currency.euroToPLN
})

export default connect(mapStateToProps)(TheBiggestTransaction)