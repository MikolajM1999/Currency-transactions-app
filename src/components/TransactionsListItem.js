import React from 'react'
import { connect } from 'react-redux'
import { removeTransaction } from '../actions/transactions'

export const TransactionListItem = ({ name, euro, id, euroToPLN, removeTransaction }) => {
   const onClick = () => {
      removeTransaction(id)
   }
   return (
      <div className="transaction-list-item-div">
         <p className="transaction-list-item-div__p">{name === '' ? 'unnamed' : name}</p>
         <p className="transaction-list-item-div__p">{Math.round(euro * 100) / 100} EURO = {Math.round(euro * euroToPLN * 100) / 100} PLN</p>
         <button className="transaction-list-item-div__button" onClick={onClick}>remove</button>
      </div>
   )
}

const mapDispatchToProps = (dispatch) => ({
   removeTransaction: (id) => dispatch(removeTransaction(id))
})

export default connect(undefined, mapDispatchToProps)(TransactionListItem)