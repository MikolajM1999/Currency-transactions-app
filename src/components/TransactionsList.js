import React from 'react'
import { connect } from 'react-redux'
import TransactionsListItem from './TransactionsListItem'

export const TransactionsList = (props) => (
   <section className="list-of-transactions app__list">
      {props.transactions.length === 0 ? (
         <div className="add-info list-of-transactions__div">
            <span className="add-info__span">Add new transactions.</span>
         </div>
      ) : (
            props.transactions.map((transaction) => {
               return <TransactionsListItem key={transaction.id} {...transaction} euroToPLN={props.euroToPLN} />
            })
         )}
   </section>
)

const mapStateToProps = (state) => ({
   euroToPLN: state.currency.euroToPLN,
   transactions: state.transactions
})

export default connect(mapStateToProps)(TransactionsList)