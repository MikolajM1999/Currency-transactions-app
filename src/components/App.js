import React from 'react'
import Header from './Header'
import CurrencyHeader from './CurrencyHeader'
import AddTransactionEuroForm from './AddTransactionEuroForm'
import TransactionsList from './TransactionsList'
import TheBiggestTransaction from './TheBiggestTransaction'

export const App = () => (
   <div className="app">
      <Header />
      <CurrencyHeader />
      <AddTransactionEuroForm />
      <div className="app__div-container-transactions">
         <TransactionsList />
         <TheBiggestTransaction />
      </div>
   </div>
)

export default App