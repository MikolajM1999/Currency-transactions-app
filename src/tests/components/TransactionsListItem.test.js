import React from 'react'
import { shallow } from 'enzyme'
import { TransactionListItem } from '../../components/TransactionsListItem'

test('should render TransactionListItem correctly with provided title', () => {
   const props = {
      name: 'AAA',
      euro: 100,
      id: '123',
      euroToPLN: 4.29,
      removeTransaction: jest.fn()
   }
   const wrapper = shallow(<TransactionListItem props={props} />)
   expect(wrapper).toMatchSnapshot()
})

test('should render TransactionListItem correctly without provided title', () => {
   const props = {
      name: '',
      euro: 2000,
      id: '123456',
      euroToPLN: 4.29,
      removeTransaction: jest.fn()
   }
   const wrapper = shallow(<TransactionListItem props={props} />)
   expect(wrapper).toMatchSnapshot()
})