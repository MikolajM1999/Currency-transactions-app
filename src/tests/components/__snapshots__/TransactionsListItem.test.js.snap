// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TransactionListItem correctly with provided title 1`] = `
<div
  className="transaction-list-item-div"
>
  <p
    className="transaction-list-item-div__p"
  />
  <p
    className="transaction-list-item-div__p"
  >
    NaN
     EURO = 
    NaN
     PLN
  </p>
  <button
    className="transaction-list-item-div__button"
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;

exports[`should render TransactionListItem correctly without provided title 1`] = `
<div
  className="transaction-list-item-div"
>
  <p
    className="transaction-list-item-div__p"
  />
  <p
    className="transaction-list-item-div__p"
  >
    NaN
     EURO = 
    NaN
     PLN
  </p>
  <button
    className="transaction-list-item-div__button"
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
